odoo/addons/base/models/ir_ui_view.py

```diff
@@ -1171,50 +1222,79 @@ def _postprocess_tag_calendar(self, node, name_manager, node_info):
+                    # To make mandatory to have both groups, wrap the field node in a <t> node with the group
+                    # set on the field in the Python model
+                    # e.g. <t groups="base.group_system"><field name="foo" groups="base.group_no_one"/></t>
+                    # The <t> node will be removed later, in _postprocess_access_rights.
+                    node_t = E.t(groups=field.groups, postprocess_added='1')
+                    node.getparent().replace(node, node_t)
+                    node_t.append(node)
+                else:
+                    node.set('groups', field.groups)
+            if (
+                node_info.get('view_type') == 'form'
+                and field.type in ('one2many', 'many2many')
+                and not node.get('widget')
+                and node.get('invisible') not in ('1', 'True')
+                and not name_manager.parent
+            ):
+                # Embed kanban/tree/form views for visible x2many fields in form views
+                # if no widget or the widget requires it.
+                # So the web client doesn't have to call `get_views` for x2many fields not embedding their view
+                # in the main form view.
+                for arch, _view in self._get_x2many_missing_view_archs(field, node, node_info):
+                    node.append(arch)
+
+            if field.relational:
+                domain = (
+                    node.get('domain')
+                    or node_info['editable'] and field._description_domain(self.env)
+                )
+                if isinstance(domain, str):
+                    vnames = get_expression_field_names(domain)
+                    name_manager.must_have_fields(node, vnames, ('domain', domain))
+            context = node.get('context')
+            if context:
+                vnames = get_expression_field_names(context)
+                name_manager.must_have_fields(node, vnames, ('context', context))
 
-            name_manager.has_field(node, node.get('name'), attrs)
+            for child in node:
+                if child.tag in ('form', 'tree', 'graph', 'kanban', 'calendar'):
+                    node_info['children'] = []
+                    sub_manager = self._postprocess_view(
+                        child, field.comodel_name, editable=node_info['editable'],
+                        parent_name_manager=name_manager, parent_field=name,
+                    )
+                    for fname, groups_uses in sub_manager.mandatory_parent_fields.items():
+                        for groups, (use, item) in groups_uses.items():
+                            name_manager.must_have_fields(item, {fname}, use, groups=groups)
+
+            if node_info['editable'] and field.type in ('many2one', 'many2many'):
+                node.set('model_access_rights', field.comodel_name)
+        elif not name.startswith(('sel_groups_', 'in_group_')):
```

Comment:
This sucks.  The crap of `res.users` now sticks on another model.  I guess we will have to live with it...